File: backend/domain/service/msg/command.go
package msg

import (
	"encoding/json"
	"main/domain/game/deck"
)

// Commands are messages sent by users to

var CommandType = struct {
	// Player buries cards
	Bury string
	// Picker calls a card (and partner)
	Call string
	// Picker chooses to go alone
	GoAlone string
	// End the current game
	EndGame string
	// Pass on the blind
	Pass string
	// Player plays a card
	PlayCard string
	// Pick up the blind
	Pick string
	// Start a new game
	StartGame string
	// User joins the game as a player
	SitDown string
	// Player leaves the game
	StandUp string
	// Updates the game settings - calling method
	UpdateCallingMethod string
	// Updates the game settings - bury resolution
	UpdateDoubleOnTheBump string
	// Updates the game settings - auto deal
	UpdateAutoDeal string
	// Updates the game settings - pick resolution
	UpdateNoPickResolution string
}{
	Bury:                   "bury",
	Call:                   "call",
	GoAlone:                "go-alone",
	EndGame:                "end-game",
	Pass:                   "pass",
	PlayCard:               "play-card",
	Pick:                   "pick",
	StartGame:              "start-game",
	SitDown:                "sit-down",
	StandUp:                "stand-up",
	UpdateCallingMethod:    "update-calling-method",
	UpdateDoubleOnTheBump:  "update-double-on-the-bump",
	UpdateAutoDeal:         "update-auto-deal",
	UpdateNoPickResolution: "update-no-pick-resolution",
}

type BuryCommandParams struct {
	Cards []*deck.Card `json:"cards"`
}

type CallCommandParams struct {
	Card *deck.Card `json:"card"`
}

type PlayCardCommandParams struct {
	Card *deck.Card `json:"card"`
}

type UpdateAutoDealParams struct {
	AutoDeal bool `json:"autoDeal"`
}

type UpdateCallingMethodParams struct {
	CallingMethod string `json:"callingMethod"`
}

type UpdateDoubleOnTheBumpParams struct {
	DoubleOnTheBump bool `json:"doubleOnTheBump"`
}

type UpdateNoPickResolutionParams struct {
	NoPickResolution string `json:"noPickResolution"`
}

func ExtractParams[T any](data interface{}) (*T, error) {
	params := new(T)
	if err := json.Unmarshal([]byte(data.(string)), params); err != nil {
		return nil, err
	}
	return params, nil
}

File: frontend/src/types/message/command.ts
import { COMMAND_TYPES } from "../../constants/message";
import { Card } from "../card";
import { CallingMethod, NoPickResolution } from "../game";
import { BaseMessage } from "./base";

export interface UpdateCallingMethodCommand extends BaseMessage {
    messageType: typeof COMMAND_TYPES.UPDATE_CALLING_METHOD;
    data: {
        callingMethod: CallingMethod | string;
    }
}

export interface UpdateNoPickResolutionCommand extends BaseMessage {
    messageType: typeof COMMAND_TYPES.UPDATE_NO_PICK_RESOLUTION;
    data: {
        noPickResolution: NoPickResolution | string;
    }
}

export interface UpdateDoubleOnTheBumpCommand extends BaseMessage {
    messageType: typeof COMMAND_TYPES.UPDATE_DOUBLE_ON_THE_BUMP;
    data: {
        doubleOnTheBump: boolean;
    }
}

export interface StartGameCommand extends BaseMessage {
    messageType: typeof COMMAND_TYPES.START_GAME;
    data: undefined;
}

export interface EndGameCommand extends BaseMessage {
    messageType: typeof COMMAND_TYPES.END_GAME;
    data: undefined;
}

export interface BuryCommand extends BaseMessage {
    messageType: typeof COMMAND_TYPES.BURY;
    data: { cards: Card[] };
}

export interface CallCommand extends BaseMessage {
    messageType: typeof COMMAND_TYPES.CALL;
    data: { card: Card };
}

export interface GoAloneCommand extends BaseMessage {
    messageType: typeof COMMAND_TYPES.GO_ALONE;
    data: undefined;
}

export interface PickCommand extends BaseMessage {
    messageType: typeof COMMAND_TYPES.PICK;
    data: undefined;
}

export interface PassCommand extends BaseMessage {
    messageType: typeof COMMAND_TYPES.PASS;
    data: undefined;
}

export interface PlayCardCommand extends BaseMessage {
    messageType: typeof COMMAND_TYPES.PLAY_CARD;
    data: { card: Card };
}

export interface CallLastHandCommand extends BaseMessage {
    messageType: typeof COMMAND_TYPES.CALL_LAST_HAND;
    data: undefined;
}


export type CommandMessageType = typeof COMMAND_TYPES[keyof typeof COMMAND_TYPES];

export type CommandMessage =
    | UpdateCallingMethodCommand
    | UpdateNoPickResolutionCommand
    | UpdateDoubleOnTheBumpCommand
    | StartGameCommand
    | EndGameCommand
    | BuryCommand
    | CallCommand
    | GoAloneCommand
    | PickCommand
    | PassCommand
    | PlayCardCommand
    | CallLastHandCommand;